store/views.py

```diff
@@ -86,21 +111,78 @@ def add_to_bag(request) -> JsonResponse:
 
-    bag: list = request.session.get('bag', [])
+    bag: list = request.session.get("bag", [])
 
     product_found = False
     for product in bag:
-        if product['product_id'] == product_id:
-            product['quantity'] += 1
+        if product["product_id"] == product_id:
+            product["quantity"] += 1
             product_found = True
             break
-    
+
     if not product_found:
-        bag.append({'product_id': product_id, 'quantity': 1})
+        bag.append({"product_id": product_id, "quantity": 1})
+
+    request.session["bag"] = bag
+
+    total_items = sum(product["quantity"] for product in bag)
+    request.session["total_items"] = total_items
 
-    request.session['bag'] = bag
+    return JsonResponse({"status": "success", "total_items": total_items})
 
-    total_items = sum(product['quantity'] for product in bag)
-    request.session['total_items'] = total_items
 
-    return JsonResponse({'status': 'success', 'total_items': total_items})
\ No newline at end of file
+class ProductAdmin(LoginRequiredMixin, ListView):
+    model = Product
+    context_object_name = "products"
+    template_name = "store/product_admin.html"
+    login_url = "/accounts/login"
+
+    def get_queryset(self) -> QuerySet[Product]:
+        # return products only from the store they own
+        return Product.objects.filter(store__owner=self.request.user)
+
+    def get_context_data(self, **kwargs: Any) -> dict[str, Any]:
+        context = super().get_context_data(**kwargs)
+
+        context["store"] = self.request.user.store_set.get()
+        return context
+
+
+@login_required
+def product_admin_modify(request, pk):
+    product = get_object_or_404(Product, pk=pk)
+    form = ProductAdminForm(instance=product)
+
+    if request.method == "POST":
+        form = ProductAdminForm(request.POST, request.FILES, instance=product)
+
+        if form.is_valid():
+            if "update" in request.POST:
+                form.save()
+                messages.add_message(
+                    request,
+                    messages.INFO,
+                    f"{product.name}({product.pk}) successfully saved.",
+                )
+            elif "delete" in request.POST:
+                product.delete()
+                messages.add_message(
+                    request,
+                    messages.INFO,
+                    f"{product.name}(ID {product.pk}) successfully deleted.",
+                )
+            return HttpResponseRedirect(reverse("store:product_admin"))
+
+    return render(request, "store/product_admin_modify.html", {"form": form})
+
+
+class ProductAdminAdd(LoginRequiredMixin, CreateView):
+    form_class = ProductAdminForm
+    template_name = "store/product_admin_add.html"
+    success_url = reverse_lazy("store:product_admin")
+
+    def form_valid(self, form):
+        product = form.save(commit=False)
+        product.store = Store.objects.get(owner=self.request.user)
```

Comment:
again here we see where a potential albeit rare issue could occur if a user creates multiple stores.
Seeing this twice now causes me to believe that creating a queryset manager could be ideal in this case such as `Store.objects.for_user()` and deal with the logic that returns the first store there. A queryset manager would be even more ideal for products.

Reply:
I love it, thanks for suggestion. I'll make use of custom managers with this logic.